---
title: "Platinum_192_XY"
author: "Xinyi"
date: "2023-02-07"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Preparation
### prepare packages 
```{r, libraries, echo=TRUE, message=FALSE}
# CRAN packages
library(ICAMS)
#BiocManager::install("ICAMS")
library(philentropy)
library(factoextra)
library(gplots)
library(nloptr)

# Bioconductor
# library(BSgenome.Hsapiens.1000genomes.hs37d5)
# If not available,
# install.packages("BiocManager")
# BiocManager::install("BSgenome.Hsapiens.1000genomes.hs37d5")

# Other
devtools::load_all(".")
1# library(mSigBG)
# if not available
# install.packages("remotes")
# remotes::install_github("steverozen/mSigBG", ref = "1.0-branch")
#library(PCAWG7)
library(BSgenome.Hsapiens.1000genomes.hs37d5)
# if not available
# remotes::install_github("steverozen/PCAWG7")
```

### Prepare SBS192 Catalogs
Load the vcf files 
```{r}
hepg2.cis_vcfs <- list.files(path = "../data/VCFs/HepG2_Cis/", full.names = TRUE)
mcf10a.cis_vcfs <- list.files(path = "../data/VCFs/MCF10A_Cis/" , full.names = TRUE)

hepg2.car_vcfs <- list.files(path = "../data/VCFs/HepG2_Carb/" , full.names = TRUE)
mcf10a.car_vcfs <- list.files(path = "../data/VCFs/MCF10A_Carb/" , full.names = TRUE)

hepg2.oxa_vcfs <- list.files(path = "../data/VCFs/HepG2_Oxa/" , full.names = TRUE)
mcf10a.oxa_vcfs <- list.files(path = "../data/VCFs/MCF10A_Oxa/" , full.names = TRUE)
```

Covert vcfs to SBS192 Catalog & Save to Rdata 
```{r}
# These vcf seems to use 38, dk why the previous R markdown uses BSgenome.Hsapiens.1000genomes.hs37d5

#HepG2

if (!file.exists("../data/hepg2.cis.spec.Rdata")){
  hepg2.cis.spec <- ICAMS::VCFsToCatalogs(hepg2.cis_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome")$catSBS192
  save(hepg2.cis.spec, file = "../data/hepg2.cis.spec.Rdata")
}else{
  load("../data/hepg2.cis.spec.Rdata")
}

if (!file.exists("../data/hepg2.car.spec.Rdata")){
  hepg2.car.spec <- ICAMS::VCFsToCatalogs(hepg2.car_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome",
               num.of.cores = 4)$catSBS192
  save(hepg2.car.spec, file = "../data/hepg2.car.spec.Rdata")
}else{
  load("../data/hepg2.car.spec.Rdata")
}

if (!file.exists("../data/hepg2.oxa.spec.Rdata")){
  hepg2.oxa.spec <- ICAMS::VCFsToCatalogs(hepg2.oxa_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome",
               num.of.cores = 4)$catSBS192
  save(hepg2.oxa.spec, file = "../data/hepg2.oxa.spec.Rdata")
}else{
  load("../data/hepg2.oxa.spec.Rdata")
}


# MCF10A
if (!file.exists("../data/mcf10a.cis.spec.Rdata")){
  mcf10a.cis.spec <- ICAMS::VCFsToCatalogs(mcf10a.cis_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome",
               num.of.cores = 4)$catSBS192
  save(mcf10a.cis.spec, file = "../data/mcf10a.cis.spec.Rdata")
}else{
  load("../data/mcf10a.cis.spec.Rdata")
}

if (!file.exists("../data/mcf10a.car.spec.Rdata")){
  mcf10a.car.spec <- ICAMS::VCFsToCatalogs(mcf10a.car_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome",
               num.of.cores = 4)$catSBS192
  save(mcf10a.car.spec, file = "../data/mcf10a.car.spec.Rdata")
}else{
  load("../data/mcf10a.car.spec.Rdata")
}

if (!file.exists("../data/mcf10a.oxa.spec.Rdata")){
  mcf10a.oxa.spec <- ICAMS::VCFsToCatalogs(mcf10a.oxa_vcfs,
               ref.genome ="hg38",
               variant.caller = "strelka",
               region = "genome",
               num.of.cores = 4)$catSBS192
  save(mcf10a.oxa.spec, file = "../data/mcf10a.oxa.spec.Rdata")
}else{
  load("../data/mcf10a.oxa.spec.Rdata")
}
```


### Functions for later
#### Catalog Plotting Functions
```{r}
pcat <- function(catalog) {
  par(pin = c(3*ncol(catalog), 1))
  par(mfrow = c(ncol(catalog), 1))
  par(mar = c(2, 4, 4, 2))
  par(cex = 0.8)
  par(cex.main = 1.4) 
  xlabels <- FALSE
  for (i in 1:ncol(catalog)) {
    if (FALSE && i == ncol(catalog)) { # FALSE an experiment
      xlabels <- TRUE
      par(mar = c(4.5, 4, 3, 2))
    }
    
    ICAMS::PlotCatalog(catalog[ , i, drop = FALSE],
                       upper = (i == 1),
                       xlabels = xlabels)
  }
}
```

#### Background separation function: Separate the background and target signatures

```{r}
one.separation <- function(sig.name, spectra, bg.sig.info, my.seed = 101010) {
  
  set.seed(my.seed, kind = "L'Ecuyer-CMRG")
  
  if (is.null(attr(spectra, "abundance"))) {
    stop("NULL abundance for ", sig.name)
  }
  
  ret <- SeparateSignatureAndSpectra(
    spectra          = spectra,
    bg.sig.info      = bg.sig.info,
    m.opts           = NULL,
    start.b.fraction = 0.5,
    sig.name         = sig.name
  )

  return(ret)
}
```

#### Two Ways of getting Inferred spectra 
##### Method 1: obs.spectra - (bg.exposure * the background signature) 
Inferred spectra obtained from observed spectra minus background spectra 
```{r}
target_sig_M1 <- function(obs.spectra, spectra.sep) {
  bg.exposure <- spectra.sep$exposures.to.bg.sig
  bg.spectra <- spectra.sep$inferred.bg.spectra
  bg.sig <- spectra.sep$inferred.bg.spectra/colSums(bg.spectra)
  
  target.spectra <- obs.spectra[,1, drop=FALSE]  - bg.exposure[1] * bg.sig[, 1, drop=FALSE]
  if(length(bg.exposure)>1){
    for (i in 2:length(bg.exposure)){
      target.spectra <- cbind(target.spectra, obs.spectra[,i, drop=FALSE]  - bg.exposure[i] * bg.sig[, i, drop=FALSE])
    }
  }
  #target.sig.method1 <- target.spectra/colSums(target.spectra) # convert clonal target spectra to signature 
  
  return(target.spectra)
  #return(target.sig.method1)
}
```

##### Method 2: (obs.counts - bg.exposure) * sig.to.return 
Inferred spectra obtained from inferred target signature multiply # of target mutations (observed -bg) 
```{r}
# No. of mutation due to bg in each sample
target_sig_M2 <- function(obs.spectra, spectra.sep){
  bg.exposures <- spectra.sep$exposures.to.bg.sig
  names(bg.exposures) <- sub("-inf.bg.spect", "", colnames(spectra.sep$inferred.bg.spectra), fixed = TRUE)
  
  target.sig <-spectra.sep$inferred.target.sig.as.catalog
  
  inferred.target.count <- sum(obs.spectra[,1]) - bg.exposures[1]
  target.spec.method2 <- inferred.target.count * target.sig
  
  if(length(bg.exposures>1)){
    for (j in 2:length(bg.exposures)){
      inferred.target.count <- sum(obs.spectra[,j]) - bg.exposures[j]
      temp.inferred.spectra <- inferred.target.count * target.sig
      target.spec.method2 <- cbind(target.spec.method2,temp.inferred.spectra)
    }
  }
  return(target.spec.method2)
}

```

### Testing whether the old and new data looks similar 
```{r}
hepg2.cis.test37 <- ICAMS::StrelkaSBSVCFFilesToCatalog(
    files      = "/home/e0261891/project_folder/mSigBG/data-raw/cisplatin-vcfs/HepG2_Cis//HepG2_Cis_cl1.vcf",
    ref.genome = BSgenome.Hsapiens.1000genomes.hs37d5,
    region     = "genome")$catSBS192
```

```{r, fig.height= 2, fig.width= 7}
pcat(hepg2.cis.spec[,1,drop=FALSE])
```

```{r, fig.height= 2, fig.width= 7}
pcat(hepg2.cis.test37)
```
```{r}
setequal(hepg2.cis.spec[,1,drop=FALSE],hepg2.cis.test37)
```

```{r}
philentropy::cosine_dist(hepg2.cis.spec[,1,drop=FALSE],hepg2.cis.test37, testNA = TRUE)
```


### Prepare Intrinsic(Background) Spectra
```{r}
# the background analysis is done on GRCh37 
hepg2.ctrl_vcfs <- list.files(path = "../data/VCFs/HepG2_Control/", full.names = TRUE)

if (!file.exists("../data/hepg2.ctrl.spec.Rdata")){
  hepg2.ctrl.spec <- ICAMS::VCFsToCatalogs(hepg2.ctrl_vcfs,
               ref.genome ="GRCh37",
               variant.caller = "strelka",
               region = "genome")$catSBS192
  save(hepg2.ctrl.spec, file = "../data/hepg2.ctrl.spec.Rdata")
}else{
  load("../data/hepg2.ctrl.spec.Rdata")
}
```

```{r}
bg.hepg2.192 <- MakeBackgroundInfo(hepg2.ctrl.spec)
```

```{r}
mcf10a.ctrl_vcfs <- list.files(path = "../data/VCFs/MCF10A_Control/", full.names = TRUE)

if (!file.exists("../data/mcf10a.ctrl.spec.Rdata")){
  mcf10a.ctrl.spec <- ICAMS::VCFsToCatalogs(mcf10a.ctrl_vcfs,
               ref.genome ="GRCh37",
               variant.caller = "strelka",
               region = "genome")$catSBS192
  save(mcf10a.ctrl.spec, file = "../data/mcf10a.ctrl.spec.Rdata")
}else{
  load("../data/mcf10a.ctrl.spec.Rdata")
}
```

```{r}
bg.mcf10a.192 <- MakeBackgroundInfo(mcf10a.ctrl.spec)
```

### Separate intrinsic and background
! RATE LIMITING STEP !
```{r}
if (!file.exists("../data/hepg2.cis.sep.Rdata")) {
  hepg2.cis.sep <- one.separation("hepg2.cis", hepg2.cis.spec, bg.hepg2.192)
  save(hepg2.cis.sep, file = "../data/hepg2.cis.sep.Rdata")
} else {
  load("../data/hepg2.cis.sep.Rdata")
}
```


```{r}
if (!file.exists("../data/mcf10a.cis.sep.Rdata")) {
  mcf10a.cis.sep <- one.separation("mcf10a.cis", mcf10a.cis.spec, bg.mcf10a.192)
  save(mcf10a.cis.sep, file = "../data/mcf10a.cis.sep.Rdata")
} else {
  load("../data/mcf10a.cis.sep.Rdata")
}
```


```{r, fig.height=2, fig.width=7}
pcat(bg.hepg2.192$background.sig)
```
```{r, fig.height=2, fig.width=7}
# hepg2.cis.sep is obtained from one_separation using bg.hepg2.192
pcat(hepg2.cis.sep$inferred.target.sig.as.catalog)
```


### Comparing!

```{r}
ts.hepg2.cis <- target_sig_M1(hepg2.cis.spec, hepg2.cis.sep)
ts.mcf10a.cis <- target_sig_M1(mcf10a.cis.spec, mcf10a.cis.sep)
```

```{r, fig.height=21, fig.width=10}
pcat(cbind(ts.hepg2.cis,ts.mcf10a.cis))
```

# Pairwise comparison for the cisplatin
comparisons within HepG2-cisplatin
```{r}
colnames <- c()
pw.subsim = matrix(data=NA, nrow=6, ncol=0)
x=1
for (i in 1:(ncol(ts.hepg2.cis)-1)){
  for (j in (i+1):ncol(ts.hepg2.cis)){
    tmp.subsim <- subcolor_sim(ts.hepg2.cis[,i,drop=FALSE],ts.hepg2.cis[,j,drop=FALSE])[["by.color"]]
    colnames <- c(colnames, paste(sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[j], fixed = TRUE),
                                  sep="-"))
    pw.subsim <- cbind(pw.subsim, tmp.subsim)
  }
}
colnames(pw.subsim) <- colnames
rownames(pw.subsim) <- names(tmp.subsim)

pw.subsim
```


```{r}

matplot(pw.subsim, type = "b",pch=1, col = 1:21, xlab = "1.C>A   2.C>G   3.C>T   4.T>A   5.T>C   6.T>G")
legend("bottomright", legend = colnames(pw.subsim), col=1:21, pch=1, cex = 0.8)
```

comparisons within MCF10A-cisplatin
```{r}
colnames <- c()
pw.subsim = matrix(data=NA, nrow=6, ncol=0)
x=1
for (i in 1:(ncol(ts.mcf10a.cis)-1)){
  for (j in (i+1):ncol(ts.mcf10a.cis)){
    tmp.subsim <- subcolor_sim(ts.mcf10a.cis[,i,drop=FALSE],ts.mcf10a.cis[,j,drop=FALSE])[["by.color"]]
    colnames <- c(colnames, paste(sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[j], fixed = TRUE),
                                  sep="-"))
    pw.subsim <- cbind(pw.subsim, tmp.subsim)
  }
}
colnames(pw.subsim) <- colnames
rownames(pw.subsim) <- names(tmp.subsim)

pw.subsim
```
```{r}
rownames(pw.subsim)
```

```{r}
matplot(pw.subsim, type = "b",pch=1,col = 1:ncol(pw.subsim), xlab = "1.C>A   2.C>G   3.C>T   4.T>A   5.T>C   6.T>G")
legend("bottomright", legend = colnames(pw.subsim), col=1:ncol(pw.subsim), pch=1, cex = 0.5)
```


comparisons across HepG2 & MCF10A cisplatin
```{r}

colnames <- c()
pw.subsim = matrix(data=NA, nrow=6, ncol=0)
x=1
for (i in 1:(ncol(ts.hepg2.cis))){
  for (j in 1:ncol(ts.mcf10a.cis)){
    tmp.subsim <- subcolor_sim(ts.hepg2.cis[,i,drop=FALSE],ts.mcf10a.cis[,j,drop=FALSE])[["by.color"]]
    colnames <- c(colnames, paste(sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[j], fixed = TRUE),
                                  sep="-"))
    pw.subsim <- cbind(pw.subsim, tmp.subsim)
  }
}
colnames(pw.subsim) <- colnames
rownames(pw.subsim) <- names(tmp.subsim)

pw.subsim

```

```{r}
matplot(pw.subsim, type = "b",pch=1,col = 1:ncol(pw.subsim), xlab = "1.C>A   2.C>G   3.C>T   4.T>A   5.T>C   6.T>G")
legend("bottomright", legend = colnames(pw.subsim), col=1:ncol(pw.subsim), pch=1, cex = 0.6, ncol=2)
```

comparisons within HepG2-cisplatin (weighted)
```{r}
pw.subsim <- c()
for (i in 1:(ncol(ts.hepg2.cis)-1)){
  for (j in (i+1):ncol(ts.hepg2.cis)){
    tmp.subsim <- subcolor_sim(ts.hepg2.cis[,i,drop=FALSE],ts.hepg2.cis[,j,drop=FALSE])[["weighted.avg"]]
    names(tmp.subsim) <- paste(sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[j], fixed = TRUE),
                                  sep="-")
    pw.subsim <- c(pw.subsim, tmp.subsim)
  }
}
pw.subsim
```

```{r}
mean(pw.subsim)
```

```{r}
plot(pw.subsim, type = "b", xaxt = "n", xlab = "HepG2-HepG2")
axis(1, at=1:length(pw.subsim), labels = gsub("HepG2_Cis_", "", names(pw.subsim), fixed = TRUE))
```


comparisons within mcf10a-cisplatin (weighted)
```{r}
pw.subsim <- c()
for (i in 1:(ncol(ts.mcf10a.cis)-1)){
  for (j in (i+1):ncol(ts.mcf10a.cis)){
    tmp.subsim <- subcolor_sim(ts.mcf10a.cis[,i,drop=FALSE],ts.mcf10a.cis[,j,drop=FALSE])[["weighted.avg"]]
    names(tmp.subsim) <- paste(sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[j], fixed = TRUE),
                                  sep="-")
    pw.subsim <- c(pw.subsim, tmp.subsim)
  }
}

pw.subsim
```

```{r}
mean(pw.subsim)
```

```{r}
plot(pw.subsim, type = "b", xaxt = "n", xlab = "MCF10A-MCF10A")
axis(1, at=1:length(pw.subsim), labels = gsub("MCF10A_Cis_", "", names(pw.subsim), fixed = TRUE))
```

comparisons within HepG2 VS Mcf10a cisplatin (weighted)
```{r}
pw.subsim <- c()
for (i in 1:ncol(ts.hepg2.cis)){
  for (j in 1:ncol(ts.mcf10a.cis)){
    tmp.subsim <- subcolor_sim(ts.hepg2.cis[,i,drop=FALSE],ts.mcf10a.cis[,j,drop=FALSE])[["weighted.avg"]]
    names(tmp.subsim) <- paste(sub("_hg38_SNVintersect", "", colnames(ts.hepg2.cis)[i], fixed = TRUE),
                                  sub("_hg38_SNVintersect", "", colnames(ts.mcf10a.cis)[j], fixed = TRUE),
                                  sep="-")
    pw.subsim <- c(pw.subsim, tmp.subsim)
  }
}

pw.subsim
```

```{r}
mean(pw.subsim)
```

```{r}
xname <- sub("HepG2_Cis_", "h", names(pw.subsim), fixed = TRUE)
xname <-sub("MCF10A_Cis_", "m" , xname, fixed = TRUE)
plot(pw.subsim, type = "b", xaxt = "n", xlab = "HepG2-MCF10A")
axis(1, at=1:length(pw.subsim), labels =xname)
```




